### PR TITLE
cmake: enforce c++14 when enabling root7

### PR DESCRIPTION
**Pull request by linev**
I have idea how to deal with webui option, but may be there is a better way.
But in principal function is there - one can disable now `root7` and/or `webui` when they are not needed

### CODE REVIEW

**Comment by linev** [comment]
@phsft-bot build with flags -Droot7=ON

**Comment by linev** [comment]
cmake works now as expected, but implementation may be not the most optimal. 
Especially `if("${webui}" STREQUAL "OFF")` statement for `webui` option.

File: cmake/modules/RootBuildOptions.cmake [hunk 1]
```diff
@@ -293,6 +293,10 @@ if(NOT CMAKE_CXX_STANDARD GREATER 11)
   set(webui_defvalue OFF)
 endif()
 
+if("${webui}" STREQUAL "OFF")
```
Comment (Axel-Naumann) [inline cmake/modules/RootBuildOptions.cmake]: `if (NOT webui AND NOT "${webui}" STREQUAL "")` will also catch `"0"`, `"NO"`, `"FALSE"` etc.

File: cmake/modules/RootBuildOptions.cmake [hunk 2]
```diff
@@ -301,6 +305,15 @@ endif()
 #---Define at moment the options with the selected default values-----------------------------
 ROOT_APPLY_OPTIONS()
 
+if(root7)
+  if(NOT CMAKE_CXX_STANDARD GREATER 11)
+      set(CMAKE_CXX_STANDARD 14 CACHE STRING "C++14 standard used with root7")
```
Comment (Axel-Naumann) [inline cmake/modules/RootBuildOptions.cmake]: I think changing the `CMAKE_CXX_STANDARD` without checking a user-provided value is bad. Probably even with checking a user-provided value. Can we error out saying "please set -DCMAKE_CXX_STANDARD=14 or higher"?

File: cmake/modules/RootBuildOptions.cmake [hunk 3]
```diff
@@ -301,6 +305,15 @@ endif()
...
+   endif()
+   if(NOT webui AND NOT webui_disabled)
+      set(webui ON CACHE BOOL "(webui build automatically when root7 is enabled)" FORCE)
```
Comment (Axel-Naumann) [inline cmake/modules/RootBuildOptions.cmake]: "build" -> "built"?

File: cmake/modules/RootBuildOptions.cmake (hunk 2 above)
Comment (linev) [reply cmake/modules/RootBuildOptions.cmake]: >  Can we error out saying "please set -DCMAKE_CXX_STANDARD=14 or higher"?

Probably my code is imperfect, but we should handle situation when only `-Droot7=ON` is specified. I guess, it is no problem to force c++14 in such case. Only when `-DCMAKE_CXX_STANDARD=11` configured by user, one should abort with error

Comment (linev) [reply cmake/modules/RootBuildOptions.cmake]: That is default value (if any) for `CMAKE_CXX_STANDARD` variable?

Comment (Axel-Naumann) [reply cmake/modules/RootBuildOptions.cmake]: >  Only when -DCMAKE_CXX_STANDARD=11 configured by user, one should abort with error

That's fine with me. Can you look into what the default value of `CMAKE_CXX_STANDARD` is? (Or whether ROOT's CMake files have set it already when getting to this point?)

Comment (linev) [reply cmake/modules/RootBuildOptions.cmake]: > Can you look into what the default value of CMAKE_CXX_STANDARD is?

I found very small info about it. This is just default value for `CXX_STANDARD`. In my environment (cmake 3.13.1, OpenSuSE Linux) it is undefined at this place.

Comment (Axel-Naumann) [reply cmake/modules/RootBuildOptions.cmake]: Thanks, then let's just error out if `CMAKE_CXX_STANDARD` is defined, else set it (with a `message`, please). Could you update the PR accordingly?

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

File: cmake/modules/RootBuildOptions.cmake [hunk 4]
```diff
@@ -301,6 +305,18 @@ endif()
 #---Define at moment the options with the selected default values-----------------------------
 ROOT_APPLY_OPTIONS()
 
+if(root7)
+  if(NOT CMAKE_CXX_STANDARD)
+      set(CMAKE_CXX_STANDARD 14 CACHE STRING "C++14 standard used with root7")
+      message(STATUS "Enable C++14 for compilation of root7 components")
```
Comment (Axel-Naumann) [inline cmake/modules/RootBuildOptions.cmake]: "Enabling" is clearer; "enable" could be a directive to the user ("re-run cmake, but then please enable...")

File: cmake/modules/RootBuildOptions.cmake [hunk 5]
```diff
@@ -301,6 +305,18 @@ endif()
...
+  elseif(NOT CMAKE_CXX_STANDARD GREATER 11)
+      message(FATAL_ERROR ">>> C++14 standard required with root7")
```
Comment (Axel-Naumann) [inline cmake/modules/RootBuildOptions.cmake]: `"At least C++14 standard required with root7"`? (*At least*)

File: cmake/modules/RootBuildOptions.cmake [hunk 6]
```diff
@@ -293,6 +293,10 @@ if(NOT CMAKE_CXX_STANDARD GREATER 11)
   set(webui_defvalue OFF)
 endif()
 
+if (NOT webui AND NOT "${webui}" STREQUAL "")
```
Comment (Axel-Naumann) [inline cmake/modules/RootBuildOptions.cmake]: Should this go to where it's used, line 315? It seems the combination of `if(NOT webui AND NOT webui_disabled)` with `if (NOT webui AND NOT "${webui}" STREQUAL "")` is mroe complex than needed - isn't this "simply"

```
if(NOT webui AND NOT "${webui}" STREQUAL "")
  set(webui ON CACHE BOOL "(webui built automatically when root7 is enabled)" FORCE)
endif()
```
i.e. do we really need `webui_disabled`?

Comment (linev) [reply cmake/modules/RootBuildOptions.cmake]: > i.e. do we really need webui_disabled?

Problem that `webui` is cmake option with own default values. 
At this place I already have no idea if user intentionally want to disable it - that's why extra variable.

Comment (Axel-Naumann) [reply cmake/modules/RootBuildOptions.cmake]: I was talking about the variable called `webui_disabled` - do we need it, or can be simply combine its `if` with the `if` below that uses `webui_disabled`?

Comment (linev) [reply cmake/modules/RootBuildOptions.cmake]: > I was talking about the variable called webui_disabled - do we need it, or can be simply combine its if with the if below that uses webui_disabled

No, it cannot be combined. When `webui_disabled` set, `root7` is not evaluated. 
I am not big expert in cmake. I will check if following is works:

```
set(webui_defvalue ${root7})
```
Does it work before `root7` option is evaluated?

Comment (linev) [reply cmake/modules/RootBuildOptions.cmake]:
```
set(webui_defvalue ${root7})
```
Seems to be it works!

**Pull request merged**